README.md
# Credit

[![Build](https://github.com/fosskers/credit/workflows/Build/badge.svg)](https://github.com/fosskers/credit/actions)
[![](https://img.shields.io/crates/v/credit.svg)](https://crates.io/crates/credit)
![AUR version](https://img.shields.io/aur/version/credit-bin)

`credit` is a fast tool for measuring Github contributions.

Use `credit` to find out:

- Who the most productive developers are in a given country.
- Who has the most Pull Requests merged to a project.
- Who engages in the most discussion in Issues and PRs.
- How long it takes maintainers to respond to and solve Issues.
- How long it takes to get PRs merged.
- If a library would be a safe long-term (i.e. maintained) dependency.

<!-- markdown-toc start - Don't edit this section. Run M-x markdown-toc-refresh-toc -->
**Table of Contents**

- [Credit](#credit)
    - [Installation](#installation)
        - [Arch Linux](#arch-linux)
        - [Cargo](#cargo)
    - [Usage](#usage)
        - [Repository Analysis](#repository-analysis)
            - [Markdown Output](#markdown-output)
            - [JSON Output](#json-output)
            - [Large Projects](#large-projects)
        - [Developer Rankings](#developer-rankings)
    - [Configuration](#configuration)
    - [FAQ](#faq)
        - [How accurate is this?](#how-accurate-is-this)
        - [Can I see commit counts too?](#can-i-see-commit-counts-too)
        - [Why do the *Median* and *Average* values differ?](#why-do-the-median-and-average-values-differ)

<!-- markdown-toc end -->

## Installation

### Arch Linux

With an AUR-compatible package manager like
[`aura`](https://github.com/fosskers/aura):

```
sudo aura -A credit-bin
```

### Cargo

```
cargo install credit
```

## Usage

To use `credit`, you'll need a Github [Personal Access
Token](https://github.com/settings/tokens) with `public_repo` permissions. [See
here](https://github.com/fosskers/active#oauth) for an additional example.

> **💡 Note:** `credit` calls the GraphQL-based Github v4 API, which has a much
> higher rate limit than the REST-based v3 API. This allows `credit` to run
> quickly and work on projects with a long development history.
>
> You can use `credit limit` to check your current API query allowance.

### Repository Analysis

#### Markdown Output

By default, `credit` outputs text to stdout that can be piped into a `.md` file
and displayed as you wish:

```
> credit repo --token=<token> rust-lang/rustfmt

# Project Report for rustfmt

## Issues

2462 issues found, 2189 of which are now closed (88.9%).

- 1899 (77.1%) of these received a response.
- 1553 (63.1%) have an official response from a repo Owner or organization Member.

Response Times (any):
- Median: 10 hours
- Average: 34 days

Response Times (official):
- Median: 13 hours
- Average: 39 days

## Pull Requests

1821 Pull Requests found, 1650 of which are now merged (90.6%).
168 have been closed without merging (9.2%).

- 1505 (82.6%) of these received a response.
- 1379 (75.7%) have an official response from a repo Owner or organization Member.

Response Times (any):
- Median: 8 hours
- Average: 2 days

Response Times (official):
- Median: 12 hours
- Average: 2 days

Time-to-Merge:
- Median: 17 hours
- Average: 3 days

## Contributors

Top 10 Commentors (Issues and PRs):
1. nrc: 2772
2. topecongiro: 1526
3. marcusklaas: 718
4. calebcartwright: 461
5. scampi: 331
6. kamalmarhubi: 120
7. rchaser53: 103
8. cassiersg: 100
9. gnzlbg: 79
10. otavio: 63

Top 10 Code Contributors (by merged PRs):
1. topecongiro: 513
2. marcusklaas: 125
3. calebcartwright: 74
4. nrc: 72
5. scampi: 64
6. rchaser53: 57
7. davidalber: 34
8. kamalmarhubi: 31
9. ayazhafiz: 28
10. sinkuu: 24
```

> **💡 Tip:** You can pass multiple repos at once to the `repo` command. The
> results will be aggregated, which can give a good view of contributions across
> an organization.

#### JSON Output

You can also output the raw results as `--json`, which could then be piped to
tools like [`jq`](https://github.com/stedolan/jq) or manipulated as you wish:

```
> credit repo --token=<token> rust-lang/rustfmt --json
```

#### Large Projects

By default, `credit` queries for Issues and Pull Requests at the same time,
which is fast and works well for most projects. For *very* large projects,
however, this can make the Github API unhappy.

If you notice `credit` failing on projects ones with many thousands of Issues
and Pull Requests, consider the `--serial` flag. This will pull Issues first,
and then Pull Requests. `--serial` allows `credit` to even work on the [Rust
compiler](https://github.com/rust-lang/rust) itself!

```
> credit repo --token=<token> rust-lang/rust --serial
```

### Developer Rankings

`credit users` can be used to determine a rough list of the most productive Open
Source programmers in a given country. This reports a similar number to the one
seen on a "Contribution Calendar", although contributions to private
repositories have been subtracted.

```
> credit users --token=<token> --location=Switzerland
```

> **💡 Note:** Due to the nature of the query made to Github, the data fetching
> will take several minutes to complete.

```
# Top 100 Open Source Contributors in Switzerland

There are currently 18518 Github users in Switzerland.

  1. oleg-nenashev (7331 contributions)
  2. cclauss (6378 contributions)
  3. dpryan79 (5604 contributions)
  4. peterpeterparker (4869 contributions)
  5. ReneNyffenegger (4722 contributions)
  6. eregon (4415 contributions)
  7. jeremytammik (3864 contributions)
  8. liufengyun (3787 contributions)
  9. swissspidy (3775 contributions)
 10. pvizeli (3706 contributions)
... and so on
```

As with `repo`, the `--json` flag can be used to output JSON data instead.

## Configuration

A configuration file can be specified at your `XDG_CONFIG_HOME`, which by
default is `$HOME/.config/credit.toml`.


```toml
# Your Github Access Token. With this set, you need not pass `--token` on the command line.
token = "abc123"
```

## FAQ

### How accurate is this?

The numbers given by `credit` are not perfect measures of developer productivity
nor maintainer responsiveness. Please use its results in good faith.

**Response Times:** Particularly in the Open Source world, volunteer developers
are under no obligation to respond in a time frame that is most convenient for
us the users.

**Merged PRs:** Without human eyes to judge a code contribution, its importance
can be difficult to measure. Some PRs are long, but do little. Some PRs are only
a single commit, but save the company. `credit` takes the stance that, over
time, with a large enough sample size, general trends of "who's doing the work"
will emerge. **Expect weird results** for one-man projects or projects that
otherwise have a long history of pushing directly to `master` without using PRs.

**User Rankings:** What is a "Top Developer" anyway? Since it is possible to
artificially inflates one's contribution numbers, `credit` uses the following
assumption to filter out false positives:

> Users with both high contribution counts and somewhat high follower counts
> must be working on something of value.

So, at first only the top 1,000 most followed developers are considered.
Afterward, other metrics are applied to arrive at a fair list of the Top 100.

### Can I see commit counts too?

Yes! Pass `--commits` to the `repo` command. Keep in mind that this requires
more data from Github, and so will take longer to complete.

### Why do the *Median* and *Average* values differ?

Given the presence of outliers in a data set, it can sometimes be more accurate
to consider the Median and not the Mean.

In the case of maintainer response times, consider a developer who usually
responds to all new Issues within 10 minutes. Then he goes on vacation, and
misses a few until his return 2 weeks later. His Average would be skewed in this
case, but the Median would remain accurate.

`credit` doesn't attempt to remove outliers, but might in the future.


CHANGELOG.md
# credit

## 1.4.0 (2021-02-05)

#### Added

- A config file can now be defined at your `XDG_CONFIG_HOME`, which by default
  is `$HOME/.config/credit.toml`. At the moment the only field is `token`:

```toml
token = "abc123"  # Your Github Access Token.
```

With this, you no longer need to pass `--token` on the command line.

#### Changed

- The dependency `reqwest` has been removed in favour of raw `curl`. This
  reduces dependency count by about 100 crates, and the final stripped binary
  size is now 1.5mb, down from about 4.5mb.

## 1.3.0 (2020-08-24)

#### Added

- A `--version` flag to display the current version of `credit`.

## 1.2.1 (2020-08-10)

#### Changed

- Swapped `isahc` for `reqwest`, which allows `credit` to be compiled with the
  MUSL target and be fully statically linked.

## 1.2.0 (2020-07-22)

#### Added

- New `users` command to produce per-country Developer Rankings.

```
> credit users --token=<token> --location=Switzerland

# Top 100 Open Source Contributors in Switzerland

There are currently 18518 Github users in Switzerland.

  1. oleg-nenashev (7331 contributions)
  2. cclauss (6378 contributions)
  3. dpryan79 (5604 contributions)
  4. peterpeterparker (4869 contributions)
  5. ReneNyffenegger (4722 contributions)
... and so on.
```

## 1.1.1 (2020-07-18)

#### Changed

- Better release profile which produces smaller binaries.

## 1.1.0 (2020-06-28)

#### Added

- New `--start` and `--end` options for `repo` that will only consider
  contributions/comments between the given dates.
- New `--commits` option for `repo`. This adds another contributor ranking: the
  number of commits that appear in merged PRs. Keep in mind that using this
  requires more data from Github, and so takes longer to complete.

## 1.0.0 (2020-06-27)

#### Added

- New `repo` command that holds the old default behaviour. Use this to analyse
  projects.
- New `limit` command for reporting the remaining Github API query allowance for
  a given token.
- New `json` command for generating a full Markdown report for JSON results
  produced by a previous run of `credit repo --json`.

#### Changed

- `credit` now uses the GraphQL-based V4 Github API. This has drastically
  improved performance and uses far less of a user's API quota upon each run.

## 0.2.0 (2020-06-22)

#### Changed

- Allow multiple repositories to be checked at the same time. Their results are
  aggregated into a single report.

## 0.1.0 (2020-06-18)

The initial release.


src/lib.rs
//! A library for measuring Github repository contributions.

mod contribs;
mod github;
mod limit;
mod repo;

// Re-export.
pub use limit::rate_limit;

use chrono::{DateTime, Utc};
use counter::Counter;
use indicatif::ProgressBar;
use itertools::Itertools;
use serde::{Deserialize, Serialize};
use std::collections::HashMap;
use std::fmt;
use std::time::Duration;

/// A nicer collated form of the data pulled from Github regarding User
/// Contributions.
#[derive(Serialize)]
pub struct UserContribs {
    pub location: String,
    pub total_users: u32,
    pub contributions: Vec<User>,
}

impl fmt::Display for UserContribs {
    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
        write!(
            f,
            "# Top 100 Open Source Contributors in {}\n\n",
            self.location
        )?;
        write!(
            f,
            "There are currently {} Github users in {}.\n\n",
            self.total_users, self.location
        )?;
        for (i, user) in self.contributions.iter().enumerate() {
            writeln!(
                f,
                "{:3}. {} ({} contributions)",
                i + 1,
                user.login,
                user.public_contributions
            )?;
        }
        Ok(())
    }
}

/// A user and their contributions.
#[derive(Serialize)]
pub struct User {
    pub login: String,
    pub name: Option<String>,
    pub public_contributions: u32,
}

impl From<contribs::UserContribs> for User {
    fn from(uc: contribs::UserContribs) -> Self {
        let public_contributions = uc.contribs();
        User {
            login: uc.login,
            name: uc.name,
            public_contributions,
        }
    }
}

/// Any type that contains a `Thread`.
trait Threaded {
    fn the_thread(&self) -> &Thread;
}

/// A Github Issue.
#[derive(Debug)]
pub struct Issue(Thread);

impl Threaded for Issue {
    fn the_thread(&self) -> &Thread {
        &self.0
    }
}

/// A Github Pull Request.
#[derive(Debug)]
pub struct PR {
    pub thread: Thread,
    pub commits: usize,
    pub merged: Option<DateTime<Utc>>,
}

impl PR {
    /// Was this Pull Request merged?
    pub fn is_merged(&self) -> bool {
        self.merged.is_some()
    }

    /// Was this Pull Request closed without merging?
    pub fn is_closed_not_merged(&self) -> bool {
        self.thread.closed.is_some() && !self.is_merged()
    }
}

impl Threaded for PR {
    fn the_thread(&self) -> &Thread {
        &self.thread
    }
}

/// A thread of conversation on Github.
///
/// This could either be associated with an Issue or a PR.
#[derive(Debug)]
pub struct Thread {
    /// Who opened the thread?
    pub author: String,
    /// When was the thread opened?
    pub posted: DateTime<Utc>,
    /// If it's already closed, when was it?
    pub closed: Option<DateTime<Utc>>,
    /// Who responded first?
    pub first_responder: Option<String>,
    /// When, if ever, was the first response?
    pub first_response: Option<DateTime<Utc>>,
    /// When, if ever, was there an "official" response?
    pub first_official_response: Option<DateTime<Utc>>,
    /// Comment counts of everyone who participated.
    pub comments: HashMap<String, usize>,
}

/// A collection of Issue and Pull Request [`Thread`](struct.Thread.html)s.
#[derive(Debug)]
pub struct Postings {
    pub issues: Vec<Issue>,
    pub prs: Vec<PR>,
}

impl Postings {
    /// Combine the results of two repository lookups.
    pub fn combine(self, other: Postings) -> Postings {
        let mut issues = self.issues;
        let mut prs = self.prs;

        issues.extend(other.issues);
        prs.extend(other.prs);

        Postings { issues, prs }
    }

    /// Consumes the `Postings` to form all the statistics.
    pub fn statistics(self) -> Statistics {
        let all_issues = self.issues.len();

        let all_closed_issues = self.issues.iter().filter(|i| i.0.closed.is_some()).count();

        let issues_with_responses = self
            .issues
            .iter()
            .filter_map(|i| i.0.first_response)
            .count();

        let issues_with_official_responses = self
            .issues
            .iter()
            .filter_map(|i| i.0.first_official_response)
            .count();

        let issue_first_resp_time = self.resp_times(|| self.issues.iter(), |i| i.0.first_response);

        let issue_official_first_resp_time =
            self.resp_times(|| self.issues.iter(), |i| i.0.first_official_response);

        let all_prs = self.prs.len();

        let prs_with_responses = self
            .prs
            .iter()
            .filter_map(|p| p.thread.first_response)
            .count();

        let prs_with_official_responses = self
            .prs
            .iter()
            .filter_map(|p| p.thread.first_official_response)
            .count();

        let pr_first_resp_time = self.resp_times(|| self.prs.iter(), |p| p.thread.first_response);

        let pr_official_first_resp_time =
            self.resp_times(|| self.prs.iter(), |p| p.thread.first_official_response);

        let prs_merged = self.prs.iter().filter(|p| p.is_merged()).count();

        let prs_closed_without_merging =
            self.prs.iter().filter(|p| p.is_closed_not_merged()).count();

        let pr_merge_time = self.resp_times(|| self.prs.iter(), |p| p.merged);

        let mut contributor_commits = HashMap::new();
        self.prs
            .iter()
            .filter(|p| p.merged.is_some())
            .for_each(|p| {
                let counter = contributor_commits
                    .entry(p.thread.author.clone())
                    .or_insert(0);
                *counter += p.commits;
            });

        let code_contributors = self
            .prs
            .iter()
            .filter(|p| p.merged.is_some())
            .map(|p| p.thread.author.clone()) // Naughty clone.
            .collect::<Counter<_>>()
            .into_map();

        let commentors = hashmap_combine(
            self.issues
                .into_iter()
                .map(|i| i.0.comments)
                .fold(HashMap::new(), hashmap_combine),
            self.prs
                .into_iter()
                .map(|p| p.thread.comments)
                .fold(HashMap::new(), hashmap_combine),
        );

        Statistics {
            commentors,
            code_contributors,
            contributor_commits,
            all_issues,
            all_closed_issues,
            issues_with_responses,
            issues_with_official_responses,
            issue_first_resp_time,
            issue_official_first_resp_time,
            all_prs,
            prs_merged,
            prs_closed_without_merging,
            prs_with_responses,
            prs_with_official_responses,
            pr_first_resp_time,
            pr_official_first_resp_time,
            pr_merge_time,
        }
    }

    /// Gather the mean/median response times in a generic way.
    fn resp_times<'a, F, G, T, A>(&self, f: F, g: G) -> Option<ResponseTimes>
    where
        F: FnOnce() -> T,
        G: Fn(&A) -> Option<DateTime<Utc>>,
        T: Iterator<Item = &'a A>,
        A: 'a + Threaded,
    {
        let resp_times: Vec<chrono::Duration> = f()
            .filter_map(|t| g(t).map(|r| r - t.the_thread().posted))
            .sorted()
            .collect();

        if resp_times.is_empty() {
            None
        } else {
            // `to_std` will error if the `Duration` is less than 0. That shouldn't
            // happen, since comments should always occur after the initial posting
            // time of the thread.
            let median = resp_times
                .get(resp_times.len() / 2)
                .and_then(|t| t.to_std().ok())?;
            let mean: i64 = resp_times.iter().map(|d| d.num_seconds()).sum();
            let mean = Duration::from_secs(mean as u64 / resp_times.len() as u64);

            Some(ResponseTimes { median, mean })
        }
    }
}

/// Statistics involving [`Thread`](struct.Thread.html) response times.
#[derive(Debug, Deserialize, Serialize)]
pub struct ResponseTimes {
    pub median: Duration,
    pub mean: Duration,
}

impl ResponseTimes {
    /// A human-friendly report of the median reponse time.
    pub fn median_time(&self) -> String {
        ResponseTimes::period(&self.median)
    }

    /// A human-friendly report of the average reponse time.
    pub fn average_time(&self) -> String {
        ResponseTimes::period(&self.mean)
    }

    fn period(duration: &Duration) -> String {
        let hours = duration.as_secs() / 3600;
        let (num, period) = if hours > 48 {
            (hours / 24, "days")
        } else if hours > 1 {
            (hours, "hours")
        } else if hours == 1 {
            (1, "hour")
        } else {
            (duration.as_secs() / 60, "minutes")
        };
        format!("{} {}", num, period)
    }
}

/// Various compiled statistics regarding contributions to a Github repository.
///
/// For the relevant fields below, an "official" response is any made by a
/// repository Owner, an organization Member, or an invited Collaborator.
#[derive(Debug, Deserialize, Serialize)]
pub struct Statistics {
    /// All issue/PR commentors.
    pub commentors: HashMap<String, usize>,
    /// All users who had PRs merged.
    pub code_contributors: HashMap<String, usize>,
    /// The commits-in-merged-PRs count for each user.
    #[serde(default)]
    pub contributor_commits: HashMap<String, usize>,
    /// The count of all issues, opened or closed.
    pub all_issues: usize,
    /// How many of the issues have been closed?
    pub all_closed_issues: usize,
    /// All issues that have been responded to in some way.
    pub issues_with_responses: usize,
    /// All issues that have been responded to "officially".
    pub issues_with_official_responses: usize,
    /// How long does it take for someone to respond to an issue?
    pub issue_first_resp_time: Option<ResponseTimes>,
    /// How long does it take for an "official" response?
    pub issue_official_first_resp_time: Option<ResponseTimes>,
    /// The count of all PRs, opened or closed.
    pub all_prs: usize,
    /// All PRs that have been responded to in some way.
    pub prs_with_responses: usize,
    /// All PRs that have been responded to officially.
    pub prs_with_official_responses: usize,
    /// How long does it take for someone to respond to a PR?
    pub pr_first_resp_time: Option<ResponseTimes>,
    /// How long does it take for an "official" response to a PR?
    pub pr_official_first_resp_time: Option<ResponseTimes>,
    /// How many PRs were merged?
    pub prs_merged: usize,
    /// The count of all PRs which were closed with being merged.
    pub prs_closed_without_merging: usize,
    /// How long does it take for PRs to be merged?
    pub pr_merge_time: Option<ResponseTimes>,
}

impl Statistics {
    pub fn report(self, repo: &str, commits: bool) -> String {
        let issues = if self.all_issues == 0 {
            "No issues found.".to_string()
        } else {
            let (any_median, any_mean) = self
                .issue_first_resp_time
                .map(|rt| (rt.median_time(), rt.average_time()))
                .unwrap_or_else(|| ("None".to_string(), "None".to_string()));

            let (official_median, official_mean) = self
                .issue_official_first_resp_time
                .map(|rt| (rt.median_time(), rt.average_time()))
                .unwrap_or_else(|| ("None".to_string(), "None".to_string()));

            format!(
                r#"
{} issues found, {} of which are now closed ({:.1}%).

- {} ({:.1}%) of these received a response.
- {} ({:.1}%) have an official response from a repo Owner or organization Member.

Response Times (any):
- Median: {}
- Average: {}

Response Times (official):
- Median: {}
- Average: {}"#,
                self.all_issues,
                self.all_closed_issues,
                percent(self.all_closed_issues, self.all_issues),
                self.issues_with_responses,
                percent(self.issues_with_responses, self.all_issues),
                self.issues_with_official_responses,
                percent(self.issues_with_official_responses, self.all_issues),
                any_median,
                any_mean,
                official_median,
                official_mean,
            )
        };

        let prs = if self.all_prs == 0 {
            "No Pull Requests found.".to_string()
        } else {
            let (any_median, any_mean) = self
                .pr_first_resp_time
                .map(|rt| (rt.median_time(), rt.average_time()))
                .unwrap_or_else(|| ("None".to_string(), "None".to_string()));

            let (official_median, official_mean) = self
                .pr_official_first_resp_time
                .map(|rt| (rt.median_time(), rt.average_time()))
                .unwrap_or_else(|| ("None".to_string(), "None".to_string()));

            let (merge_median, merge_mean) = self
                .pr_merge_time
                .map(|rt| (rt.median_time(), rt.average_time()))
                .unwrap_or_else(|| ("None".to_string(), "None".to_string()));

            format!(
                r#"
{} Pull Requests found, {} of which are now merged ({:.1}%).
{} have been closed without merging ({:.1}%).

- {} ({:.1}%) of these received a response.
- {} ({:.1}%) have an official response from a repo Owner or organization Member.

Response Times (any):
- Median: {}
- Average: {}

Response Times (official):
- Median: {}
- Average: {}

Time-to-Merge:
- Median: {}
- Average: {}"#,
                self.all_prs,
                self.prs_merged,
                percent(self.prs_merged, self.all_prs),
                self.prs_closed_without_merging,
                percent(self.prs_closed_without_merging, self.all_prs),
                self.prs_with_responses,
                percent(self.prs_with_responses, self.all_prs),
                self.prs_with_official_responses,
                percent(self.prs_with_official_responses, self.all_prs),
                any_median,
                any_mean,
                official_median,
                official_mean,
                merge_median,
                merge_mean,
            )
        };

        let contributors = format!(
            r#"
Top 10 Commentors (Issues and PRs):
{}

Top 10 Code Contributors (by merged PRs):
{}
"#,
            self.commentors
                .into_iter()
                .sorted_by(|a, b| b.1.cmp(&a.1))
                .take(10)
                .enumerate()
                .map(|(i, (name, issues))| format!("{:2}. {}: {}", i + 1, name, issues))
                .join("\n"),
            self.code_contributors
                .into_iter()
                .sorted_by(|a, b| b.1.cmp(&a.1))
                .take(10)
                .enumerate()
                .map(|(i, (name, prs))| format!("{:2}. {}: {}", i + 1, name, prs))
                .join("\n"),
        );

        let contributor_commits = if commits {
            format!(
                r#"
Top 10 Code Contributors (by commits-in-merged-PRs):
{}
"#,
                self.contributor_commits
                    .into_iter()
                    .sorted_by(|a, b| b.1.cmp(&a.1))
                    .take(10)
                    .enumerate()
                    .map(|(i, (name, commits))| format!("{:2}. {}: {}", i + 1, name, commits))
                    .join("\n"),
            )
        } else {
            "".to_string()
        };

        format!(
            r#"# Project Report for {}

## Issues
{}

## Pull Requests
{}

## Contributors
{}{}"#,
            repo, issues, prs, contributors, contributor_commits
        )
    }
}

/// Given a repository name, look up the [`Thread`](struct.Thread.html)
/// statistics of all its Issues.
pub fn repo_threads(
    token: &str,
    ipb: &ProgressBar,
    ppb: &ProgressBar,
    serial: bool,
    commits: bool,
    start: &Option<DateTime<Utc>>,
    end: &Option<DateTime<Utc>>,
    owner: &str,
    repo: &str,
) -> anyhow::Result<Postings> {
    let i_msg = format!("Fetching Issues for {}/{}...", owner, repo);
    let p_msg = format!("Fetching Pull Requests for {}/{}...", owner, repo);

    let get_issues = || all_issues(token, start, end, owner, repo);
    let get_prs = || all_prs(token, start, end, commits, owner, repo);

    // Too much parallelism can trigger Github's abuse detection, so we offer
    // the "serial" option here.
    let (issues, prs) = if serial {
        let issues = with_progress(ipb, &i_msg, get_issues);
        let prs = with_progress(ppb, &p_msg, get_prs);
        (issues, prs)
    } else {
        rayon::join(
            || with_progress(ipb, &i_msg, get_issues),
            || with_progress(ppb, &p_msg, get_prs),
        )
    };

    Ok(Postings {
        issues: issues?,
        prs: prs?,
    })
}

/// Perform some action with an associated `ProgressBar`.
fn with_progress<F, A>(progress: &ProgressBar, msg: &str, f: F) -> A
where
    F: FnOnce() -> A,
{
    progress.enable_steady_tick(120);
    progress.set_message(&msg);
    let result = f();
    progress.finish_and_clear();
    result
}

fn all_issues(
    token: &str,
    start: &Option<DateTime<Utc>>,
    end: &Option<DateTime<Utc>>,
    owner: &str,
    repo: &str,
) -> anyhow::Result<Vec<Issue>> {
    repo::issues(token, end, &repo::Mode::Issues, owner, repo).map(|is| {
        is.into_iter()
            .filter(|i| {
                let after = start.map(|s| i.created_at >= s).unwrap_or(true);
                let before = end.map(|e| i.created_at <= e).unwrap_or(true);
                after && before
            })
            .map(|i| Issue(issue_thread(i)))
            .collect()
    })
}

fn all_prs(
    token: &str,
    start: &Option<DateTime<Utc>>,
    end: &Option<DateTime<Utc>>,
    commits: bool,
    owner: &str,
    repo: &str,
) -> anyhow::Result<Vec<PR>> {
    let mode = if commits {
        repo::Mode::PRsWithCommits
    } else {
        repo::Mode::PRs
    };
    repo::issues(token, end, &mode, owner, repo).map(|is| {
        is.into_iter()
            .filter(|i| {
                let after = start.map(|s| i.created_at >= s).unwrap_or(true);
                let before = end.map(|e| i.created_at <= e).unwrap_or(true);
                after && before
            })
            .map(|i| {
                let merged = i.merged_at;
                let commits = i.commits.as_ref().map(|cc| cc.total_count).unwrap_or(0);
                let thread = issue_thread(i);
                PR {
                    thread,
                    merged,
                    commits,
                }
            })
            .collect()
    })
}

fn ghost(author: &Option<repo::Author>) -> String {
    author
        .as_ref()
        .map(|a| a.login.clone())
        .unwrap_or_else(|| "@ghost".to_string())
}

fn issue_thread(issue: repo::Issue) -> Thread {
    let comments: Vec<repo::Comment> = issue.comments.edges.into_iter().map(|n| n.node).collect();

    // Need to be careful, since the first physical response might have been
    // from the Issue author.
    let first_comment = comments.iter().find(|c| !c.author_association.is_author());
    let first_responder = first_comment.map(|c| ghost(&c.author));
    let first_response = first_comment.map(|c| c.created_at);
    let first_official_response = comments
        .iter()
        .find(|c| c.author_association.is_official())
        .map(|c| c.created_at);
    let comment_counts = comments
        .iter()
        .map(|c| ghost(&c.author))
        .collect::<Counter<_>>()
        .into_map();

    Thread {
        author: ghost(&issue.author),
        posted: issue.created_at,
        closed: issue.closed_at,
        first_responder,
        first_response,
        first_official_response,
        comments: comment_counts,
    }
}

/// A curated list of the Top 100 users in a given location, ranked via their
/// contribution counts and weighted by followers.
pub fn user_contributions(token: &str, location: &str) -> anyhow::Result<UserContribs> {
    let total_users = contribs::user_count(token, location)?.user_count;
    let contributions = contribs::user_contributions(token, location)?
        .into_iter()
        .sorted_by(|a, b| b.contribs().cmp(&a.contribs()))
        .take(500)
        .sorted_by(|a, b| b.followers.total_count.cmp(&a.followers.total_count))
        .take(250)
        .sorted_by(|a, b| b.contribs().cmp(&a.contribs()))
        .take(100)
        .map(|uc| User::from(uc))
        .collect();

    Ok(UserContribs {
        location: location.to_string(),
        total_users,
        contributions,
    })
}

fn hashmap_combine<K, V>(mut a: HashMap<K, V>, b: HashMap<K, V>) -> HashMap<K, V>
where
    K: Eq + std::hash::Hash,
    V: std::ops::AddAssign + Copy,
{
    for (k, v) in b {
        a.entry(k).and_modify(|e| *e += v).or_insert(v);
    }

    a
}

fn percent(a: usize, b: usize) -> f64 {
    100.0 * (a as f64) / (b as f64)
}

#[test]
fn hashmap_extend() {
    let mut first = HashMap::new();
    let mut second = HashMap::new();
    let message = "ABCDEF";

    for c in message.chars() {
        first.insert(c, 1);
        second.insert(c, 1);
    }

    let third = hashmap_combine(first, second);
    let elems: Vec<usize> = third.values().map(|v| *v).collect();

    assert_eq!(vec![2, 2, 2, 2, 2, 2], elems);
}
